packages/core/agent/src/launcher/runner/launchctl-runner.ts

```diff
@@ -0,0 +1,102 @@
+
+const execPromise = util.promisify(exec);
+
+export class LaunchctlRunner implements Runner {
+  async start({ profile, logFile, errFile, daemonOptions }: RunnerStartOptions): Promise<void> {
+    try {
+      const defaultTemplatePath = path.join(
+        path.dirname(pkgUp.sync({ cwd: __dirname }) ?? raise(new Error('Could not find package.json'))),
+        PLIST_TEMPLATE_FILE,
+      );
+
+      const plistTemplate = await readFile(defaultTemplatePath, 'utf-8');
+
+      const service = this._getServiceName(profile);
+      const plistPath = await this._getPlistPath(service);
+
+      const options = [''];
+      if (daemonOptions?.metrics) {
+        options.push('<string>--metrics</string>');
+      }
+      if (daemonOptions?.ws) {
+        options.push(`<string>--ws=${daemonOptions.ws}</string>`);
+      }
+      if (daemonOptions?.config) {
+        options.push(`<string>--config=${daemonOptions.config}</string>`);
+      }
+
+      const plistContent = plistTemplate
+        .replace('{{DX_PATH}}', process.argv[1])
+        .replace('{{NODE_PATH}}', path.dirname(process.execPath))
+        .replace('{{ERROR_LOG}}', errFile)
+        .replace('{{OUT_LOG}}', logFile)
+        .replace('{{OPTIONS}}', options.length > 1 ? options.join('\n') : '')
+        .replaceAll('{{PROFILE}}', profile);
+
+      // Write the plist content to the file.
+      await writeFile(plistPath, plistContent, 'utf-8');
+
+      // Load the plist file using launchctl.
+      await execPromise(`launchctl load ${plistPath}`);
+
+      // Ensure the service is started.
+      await execPromise(`launchctl start ${service}`);
+    } catch (err: any) {
+      throw new Error(`Failed to load plist: ${err.message}`);
+    }
+  }
+
+  async stop(profile: string, force: boolean = false): Promise<void> {
+    try {
+      const service = this._getServiceName(profile);
+      const plistPath = await this._getPlistPath(service);
+
+      // Unload the plist file using launchctl.
+      await execPromise(`launchctl unload ${plistPath}`);
+
+      // Remove the plist file.
+      await unlink(plistPath);
+    } catch (err: any) {
+      throw new Error(`Failed to unload plist: ${err.message}`);
+    }
+  }
+
+  async isRunning(profile: string): Promise<boolean> {
+    try {
+      const service = this._getServiceName(profile);
+      const { stdout } = await execPromise('launchctl list');
+      return stdout.includes(service);
+    } catch (err: any) {
+      throw new Error(`Failed to check if the service is running: ${err.message}`);
+    }
+  }
+
+  private async _getPlistPath(service: string): Promise<string> {
+    const agentsPath = `${os.homedir()}/Library/LaunchAgents`;
+    if (!existsSync(agentsPath)) {
+      await mkdir(agentsPath, { recursive: true });
+    }
+    return `${agentsPath}/${service}.plist`;
+  }
+
+  private _getServiceName(profile: string): string {
+    return `org.dxos.agent.${profile}`;
```

Comment:
Extract const. What standard are we using here? (Elsewhere dxos.org/service/xxx?)

Reply:
This is the macos specific format for services, see
https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html
or https://ss64.com/osx/launchctl.html